File: Day 2 - Easy (Level 1)/LoveCalculator/LoveCalculator.cs
// BEGIN CUT HERE

// END CUT HERE
using System;
using System.Text;
using System.Text.RegularExpressions;
using System.Collections;

public class LoveCalculator {
    public string findBoy(string girl, string[] boys) {
        string res;
        return res;
    }

// BEGIN CUT HERE
    public static void Main(String[] args) {
        try  {
            eq(0,(new LoveCalculator()).findBoy("LOVE", new string[] {"JACOB","FRANK","DANO"}),"FRANK");
            eq(1,(new LoveCalculator()).findBoy("JANE", new string[] {"THOMAS","MICHAEL","INDY","LIU"}),"INDY");
            eq(2,(new LoveCalculator()).findBoy("LILLY", new string[] {"PIERRE"}),"PIERRE");
            eq(3,(new LoveCalculator()).findBoy("MERYLOV", new string[] {"JOHN","DAVE","STEVE","JOHN","DAVE"}),"DAVE");
            eq(4,(new LoveCalculator()).findBoy("LLOL", new string[] {"BVERON","CVERON","AVERON","DVERON"}),"AVERON");
            eq(5,(new LoveCalculator()).findBoy("VELYLEOCEVE", new string[] {"YVXHOVE","LCOKO","OGWSJVEVEDLE","WGFVSJEL","VLOLUVCBLLQVESWHEEKC"}),"VLOLUVCBLLQVESWHEEKC");
        } 
        catch( Exception exx)  {
            System.Console.WriteLine(exx);
            System.Console.WriteLine( exx.StackTrace);
        }
    }
    private static void eq( int n, object have, object need) {
        if( eq( have, need ) ) {
            Console.WriteLine( "Case "+n+" passed." );
        } else {
            Console.Write( "Case "+n+" failed: expected " );
            print( need );
            Console.Write( ", received " );
            print( have );
            Console.WriteLine();
        }
    }
    private static void eq( int n, Array have, Array need) {
        if( have == null || have.Length != need.Length ) {
            Console.WriteLine("Case "+n+" failed: returned "+have.Length+" elements; expected "+need.Length+" elements.");
            print( have );
            print( need );
            return;
        }
        for( int i= 0; i < have.Length; i++ ) {
            if( ! eq( have.GetValue(i), need.GetValue(i) ) ) {
                Console.WriteLine( "Case "+n+" failed. Expected and returned array differ in position "+i );
                print( have );
                print( need );
                return;
            }
        }
        Console.WriteLine("Case "+n+" passed.");
    }
    private static bool eq( object a, object b ) {
        if ( a is double && b is double ) {
            return Math.Abs((double)a-(double)b) < 1E-9;
        } else {
            return a!=null && b!=null && a.Equals(b);
        }
    }
    private static void print( object a ) {
        if ( a is string ) {
            Console.Write("\"{0}\"", a);
        } else if ( a is long ) {
            Console.Write("{0}L", a);
        } else {
            Console.Write(a);
        }
    }
    private static void print( Array a ) {
        if ( a == null) {
            Console.WriteLine("<NULL>");
        }
        Console.Write('{');
        for ( int i= 0; i < a.Length; i++ ) {
            print( a.GetValue(i) );
            if( i != a.Length-1 ) {
                Console.Write(", ");
            }
        }
        Console.WriteLine( '}' );
    }
// END CUT HERE
}

File: Day 2 - Easy (Level 1)/LoveCalculator/Solution/LoveCalculator.js
var test_case1 = {

  girl : "LOVE" ,
  boys : ["JACOB","FRANK","DANO"],
  expected : "FRANK"

},

test_case2 = {

  girl : "JANE" ,
  boys : ["THOMAS","MICHAEL","INDY","LIU"],
  expected : "INDY"

}

test_case3 = {

  girl : "LILLY" ,
  boys : ["PIERRE"],
  expected : "PIERRE"

}

test_case4 = {

  girl : "MERLYLOV" ,
  boys : ["JOHN","DAVE","STEVE","JOHN","DAVE"],
  expected : "JOHN"

}
test_case5 = {

  girl : "LLOL" ,
  boys : ["BVERON","CVERON","AVERON","DVERON"],
  expected : "AVERON"

}



console.log("\n Case 1, Expected Result: "+test_case1.expected+", Output: "+findBoy(test_case1.girl, test_case1.boys));
console.log("\n Case 2, Expected Result: "+test_case2.expected+", Output: "+findBoy(test_case2.girl, test_case2.boys));
console.log("\n Case 3, Expected Result: "+test_case3.expected+", Output: "+findBoy(test_case3.girl, test_case3.boys));
console.log("\n Case 4, Expected Result: "+test_case4.expected+", Output: "+findBoy(test_case4.girl, test_case4.boys));
console.log("\n Case 4, Expected Result: "+test_case5.expected+", Output: "+findBoy(test_case5.girl, test_case5.boys));


function findBoy(girl, boys) {
  boys.sort(); 
  
  var max = -2147483648; 
  var minb = null; 
  
  for (var i = 0; i < boys.length; i++) { 
    var score = calc(girl, boys[i]); //get calculation for each boy for comaprision
    if(score > max){ 
      max = score; 
      minb = boys[i]; 
    } 
  } 
      return minb; 
} 

function calc(girl, boy) { 
  var L,O,V,E; 
  
  //Counting for words LOVE in each boy and girl's name
  L = count(girl,'L') + count(boy,'L'); 
  O = count(girl,'O') + count(boy,'O'); 
  V = count(girl,'V') + count(boy,'V'); 
  E = count(girl,'E') + count(boy,'E'); 
  
  return ((L+O)*(L+V)*(L+E)*(O+V)*(O+E)*(V+E))%100; 
}

//Function to count occurence of charchter in string
function count(s, c) { 
  var count = 0; 
  
  for (var i = 0; i < s.length; i++) { 
    if(s.charAt(i) == c){ 
      count++; 
    } 
  } 
    return count; 
} 
